ACCUMULO-3645 Run iterators at major compaction on empty tablets

denine99 opened the pull request:
Putting this out for review to see if I am on the right track; I don't expect it to be perfect on the first shot.  I'd like to redo this for 1.6.3 if it looks good.  Open to suggestions on design and on testing compacting empty tablets.

Removed/modified lines in Tablet that stop construction of the SKVI stack when a tablet has no data.
Includes an iterator that generates data, for testing purposes. Includes one integration test. Not sure about MiniAccumulo.

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/compaction/EverythingCompactionStrategy.java:
```diff
@@ -27,7 +27,7 @@
   @Override
   public boolean shouldCompact(MajorCompactionRequest request) throws IOException {
-    return request.getFiles().size() > 0;
+      return true; // ACCUMULO-3645 compact for empty files too
```
Rather than change the default compaction strategy, it'd probably be better to introduce a new compaction strategy, which returns true, instead of implementing the optimized behavior.

denine99 replied:
The only time this will make a difference is when the USER manually issues a compaction with at least one iterator, in which case the user should expect the compaction occurs by default.

I will add a condition to the Tablet code to keep the "optimized" behavior for empty tablets on user-issued compactions that do not have any iterators configured.

joshelser replied:
I think I'd only be in favor of making the change for 1.6.3 if it's done as a new CompactionStrategy. I'd need convincing to change the default implementation in a bug fix release without seriously enumerating the corner cases.

ctubbsii replied:
A point to consider: unless it's definitely a bug fix, and we're sure we want to change the behavior to fix that bug, our versioning rules would disallow new functionality to be added in 1.6.3. It'd have to go in 1.7.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/iterators/HardListIterator.java:
```diff
@@ -0,0 +1,109 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.accumulo.core.iterators;
```
This should be in the "users" sub-package.

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:
```diff
@@ -1861,8 +1861,10 @@ private CompactionStats _majorCompact(MajorCompactionReason reason) throws IOExc
           // no work to do
-          lastCompactID = compactionId.getFirst();
-          updateCompactionID = true;
+//          lastCompactID = compactionId.getFirst();
```
What was your intention in leaving these commented out?
I'm not too familiar with how the compactionId is used, but it seems like getting this right is tricky.

denine99 replied:
Yes, touching `lastCompactID` is sensitive because Accumulo uses that to determine when a compaction has already occurred. I think the modifications are correct, but I retained the commented lines because I feel unsure.  I will remove the commented lines and try to test further.

ctubbsii replied:
I'd definitely feel more comfortable if @keith-turner reviewed it. He's the last person I know who fixed a bug in this code.

keith-turner replied:
I will look at this PR Mon.

denine99 commented:
Anyone have additional test cases to suggest?  I covered 

* multi-tablet full major compactions, 
* multi-tablet non-full major compactions, 
* canceled major compactions

denine99 commented:
Not sure why the asfbot Jenkins CI thingy found a problem.  The build, unit tests and the integration tests I wrote work for me, both on MiniAccumulo and standalone.

joshelser commented:
Checkstyle violation: src/test/java/org/apache/accumulo/test/TableOperationsIT.java[299] (regexp) RegexpSinglelineJava: Line has trailing whitespace.

https://builds.apache.org/job/Accumulo-Pull-Requests/13/console

If you're only running `mvn package`, checkstyle (and findbugs) won't be invoked. You can either run them manually, or do a `mvn verify -DskipITs` (assuming you otherwise want it to function as `mvn package` does).

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:
```diff
@@ -2548,7 +2565,8 @@ public void compactAll(long compactionId, UserCompactionConfig compactionConfig)
     synchronized (this) {
-      if (lastCompactID >= compactionId)
+      // ACCUMULO-3645 compaction may generate results if there is a compaction-scope iterator
+      if (lastCompactID >= compactionId && compactionConfig.getIterators().isEmpty())
```
If lastCompactID >= compactionId then should not do anything.  When master receives compaction request it increments counter in zookeeper.   When a tablet compacts, it records the compaction id in metadata.   Master keeps asking tablets to compact until all tablets are counter are >= the zookeeper value.   This change could lead to tablet compacting multiple times for a single user compaction request with iters.

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:
```diff
@@ -2567,11 +2585,13 @@ public void compactAll(long compactionId, UserCompactionConfig compactionConfig)
         lastCompactID = compactionId;
-      } else {
+      }
+// else
```
Should not update compaction id in metadata table and initiate compaction

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:
```diff
@@ -2022,6 +2029,16 @@ protected AccumuloConfiguration createTableConfiguration(TableConfiguration base
 
+    // short-circuit; also handles zero files case
+    if (filesToCompact.size() <= maxFilesToCompact) {
+      Set<FileRef> smallestFiles = new HashSet<FileRef>(filesToCompact.size());
```
could do

```java
      Set<FileRef> smallestFiles = new HashSet<FileRef>(filesToCompact.keySet());
      filesToCompact.clear();
      return smallestFiles;
```

slightly easier to read than for loop

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:
```diff
@@ -1913,20 +1918,22 @@ private CompactionStats _majorCompact(MajorCompactionReason reason) throws IOExc
           synchronized (this) {
-            if (lastCompactID >= compactionId.getFirst())
+            // ACCUMULO-3645 run if there are compaction iterators that may generate entries
+            if (lastCompactID >= compactionId.getFirst() && compactionIterators.isEmpty())
```
if the user req compaction is already done then should not continue

denine99 commented:
Thanks @keith-turner.  I will revisit the compaction IDs and get back with another commit.  

Do you know what the expected behavior is when we use a BadIterator in a user compaction?

denine99 commented:
I fixed up the patch from Keith's comments. This one looks good.

My mistake starting out with the patch was trying to change *every* location where an empty tablet could be disregarded during compaction.  Now the current patch has the *minimal* changes necessary to make empty tablets compact.  The biggest one is the change to the default compaction strategy.  As before, I'm open for ideas to test further.  I think we're closing in on good code.

To answer the question I posed previously, the expected behavior when a major compaction iterator throws and exception (e.g. [BadIterator](https://accumulo.apache.org/1.6/apidocs/org/apache/accumulo/test/functional/BadIterator.html)) 
is that the tablet server will fail the compaction, not updating the metadata table, and then retry the compaction.  If the error is permanent, then the compaction will fail again and again continuously.  The only way to stop the cycle is to cancel the compaction manually.  This patch maintains this behavior.

Rebased the commits to master.  Will squash once given a green light.

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:
```diff
@@ -2566,12 +2579,7 @@ public void compactAll(long compactionId, UserCompactionConfig compactionConfig)
-        updateMetadata = true;
-        lastCompactID = compactionId;
-      } else {
+      {
```
Why leave this undconditional code block?  Did a var name conflict?

denine99 replied:
No reason, except to make a smaller diff.  Removing...

keith-turner commented on core/src/main/java/org/apache/accumulo/core/iterators/user/HardListIterator.java:
```diff
@@ -0,0 +1,111 @@
+  private Range seekRng;
+
+  @Override
+  public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
```
Curious what `env.isFullMajorCompaction()` returns in the case where the tablet has no files.  Should probably return true.  Could have a test for this

denine99 replied:
I confirmed that env.isFullMajorCompaction() is true when the tablet has no files on a standalone single-node instance.  

Writing a test for it is possible by writing an iterator, similar to HardListIterator, that emits an entry with a boolean value-- the result of env.isFullMajorCompaction().  It does not seem worth it to include a new iterator class in the source code for one test like this.

keith-turner replied:
Your comment made me realize HardListIterator was in core/src/main/java,
should probably be in core/src/test/java.
On Mar 27, 2015 7:33 PM, "Dylan Hutchison" <notifications@github.com> wrote:

> In
> core/src/main/java/org/apache/accumulo/core/iterators/user/HardListIterator.java
> <https://github.com/apache/accumulo/pull/22#discussion_r27338487>:
>
> > +public class HardListIterator implements SortedKeyValueIterator<Key,Value> {
> > +  private static final Logger log = Logger.getLogger(HardListIterator.class);
> > +  public final static SortedMap<Key,Value> allEntriesToInject;
> > +  static {
> > +    SortedMap<Key,Value> t = new TreeMap<>();
> > +    t.put(new Key(new Text("a1"), new Text("colF3"), new Text("colQ3"), System.currentTimeMillis()), new Value("1".getBytes()));
> > +    t.put(new Key(new Text("c1"), new Text("colF3"), new Text("colQ3"), System.currentTimeMillis()), new Value("1".getBytes()));
> > +    t.put(new Key(new Text("m1"), new Text("colF3"), new Text("colQ3"), System.currentTimeMillis()), new Value("1".getBytes()));
> > +    allEntriesToInject = Collections.unmodifiableSortedMap(t); // for safety
> > +  }
> > +
> > +  private PeekingIterator<Map.Entry<Key,Value>> inner;
> > +  private Range seekRng;
> > +
> > +  @Override
> > +  public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
>
> I confirmed that env.isFullMajorCompaction() is true when the tablet has
> no files on a standalone single-node instance.
>
> Writing a test for it is possible by writing an iterator, similar to
> HardListIterator, that emits an entry with a boolean value-- the result of
> env.isFullMajorCompaction(). It does not seem worth it to include a new
> iterator class in the source code for one test like this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/accumulo/pull/22/files#r27338487>.
>

denine99 replied:
I tried putting HardListIterator in core/src/test/java when I first wrote it, but then it does not compile (error: cannot find symbol HardListIterator).  I think it needs to be on the main compile path to run in a live instance of Accumulo, not the testCompile path.

joshelser replied:
Yes, putting HardListIterator in src/test/java will place it in the test jar instead of the "regular" jar. I don't think we put the core test-jar on the classpath for ITs. You can (should, IMO) put it in test/src/test/java instead of core/src/main/java and that will work for your integration test.

denine99 replied:
Forgot to comment on this after committing-- I moved the class into the main code of the test project.  I did this so that I could deploy the test code JAR in a standalone Accumulo cluster after calling `mvn package -DskipITs` and then do the test.  Maybe the Accumulo build has an automated way to deploy JARs containing test src code in standalone instances via the configuration?  If not, this is a good compromise.

keith-turner commented on test/src/test/java/org/apache/accumulo/test/TableOperationsIT.java:
```diff
@@ -229,4 +239,122 @@ public void createMergeClonedTable() throws Exception {
     return map;
   }
 
+  @Test
```
Nice tests!!!!!!!!!!

denine99 replied:
Thanks =D

joshelser commented on server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:
```diff
@@ -1860,9 +1860,14 @@ private CompactionStats _majorCompact(MajorCompactionReason reason) throws IOExc
+            updateCompactionID = true;
+          } else {
+            log.debug("Major compaction by USER on 0 input files with iterators.");
+            filesToCompact = new HashMap<FileRef,DataFileValue>();
```
nit, you can just use the diamond operator

denine99 replied:
We're at Java 7 then?  Sounds good.

joshelser replied:
Yup, it looks like we moved this fixVersion to 1.7.0 which is on Java7. Accumulo 1.6 and earlier is still on java6

denine99 replied:
Hey Josh, question on open-source contributing.  There are many, many spots throughout the Tablet class where we can delete the contents of the diamond.  However, deleting them all isn't relevant to the goal of this patch: making iterators run on empty user compactions. If I deleted them all, then reviewers may find the patch confusing.

Is opening a new ticket and deleting all the diamonds the thing to do?  It seems kinda silly to use a ticket for that.  Actually if we need to open a ticket for every little issue, then no one would ever fix (for example) a typo in documentation because the overhead of a ticket is overkill.  How does it work?

joshelser replied:
No, you're actually right. It's best to keep changes as concise as you can (don't change a bunch of other things in one JIRA issue). It helps people go back in the future and understand what changes were made in relation to the JIRA issue and related conversation.

I don't recall if we've had discussion on the subject of Java7 features aside than "use the new features in new code you write as you'd like." While we support 1.5 and 1.6, if we change 1.7 to use diamond everywhere, it's likely to just cause more of a headache in merges than it's worth in the codebase. That's just my $0.02 though. Feel free to poll the dev list or make a JIRA issue about it though. Others might have different opinions :)

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/compaction/EverythingCompactionStrategy.java:
```diff
@@ -27,7 +27,7 @@
 
   @Override
   public boolean shouldCompact(MajorCompactionRequest request) throws IOException {
```
`MajorCompactionRequest` could be modified to add a `getCompactionIterators()` method, then this would have all of the info needed to make a decision.  This might clean the code up a little bit in other places.  This is just something I thought of while looking at this, not something I think has to be done.

keith-turner commented:
+1

I think this change looks good.

keith-turner commented:
@denine99 I can merge this.  Would you mind if I squished it down to one commit (I would maintain you as the author in git)?

keith-turner commented:
If we do not squish down to one commit, we would need to add issue number to all of the commit messages.

joshelser commented:
@keith-turner you can also manually set the author on the commit if you just want to do it yourself via the `--author=<author>` option on `git-commit` (`git checkout`, `git rebase -i`, and `git commit --amend --author=...`). I've done this for a few patches, didn't know if you knew about it.

keith-turner commented:
I was aware of how to change the author, I just wanted to make sure @denine99  was ok w/ me squishing the commits into one commit

denine99 commented:
@keith-turner,
  Since this is my first 'real' pull request and I could use some practice, I went ahead and squashed into one commit and rebased at master 85d254e12.  How's this look?
